Add CNYUSD as price identifiers

jishankai opened the pull request:
The DVM currently does not support the CNY/USD or USD/CNY indexes.

Supporting the CNYUSD price identifier would enable the creation of a Chinese Yuan FX derivative, backed by USD. Token minters could go short on the CNY/USD index, while token holders could go long or use synthetic fxCNY for functional purposes.

smb2796 commented on UMIPs/umip-28.md:
```diff
@@ -0,0 +1,74 @@
+We tried exchanges such as Binance, Gate and KuCoin. All of their apis don't support USDCNY.
+
+We also tried CoinMarketCap which standard plan above support historical search. It's not a good choice because it needs be paid.
+
```
Was there any consideration on using a median rather than an average?

jishankai replied:
I preferred a median to an average and I saw most of the UMIPs used a median. However, I only found two API endpoints that support the USDT/CNY historical search. I think it's not possible to use a median if you only have two numbers.

smb2796 replied:
Absolutely right. Average seems fine to me if there are only two available sources

smb2796 commented on UMIPs/umip-28.md:
```diff
@@ -0,0 +1,74 @@
+- Dispute timestamp rounding: down
+
+
+## Rationale
```
I'm curious if CNY/USD rate data was considered. Seems like there would be quite a few sources for that as opposed to CNY/USDT. Like the idea of keeping it cryptonative, but just interested in the thought process!

jishankai replied:
Sure, there should be more sources for USD/CNY than USDT/CNY. I tried to find some but I failed when I found that I can only find two sources for USDT/CNY. I'm not familiar with finance APIs and tried the first two pages of Google search 'USD/CNY' results but some aren't free such as Bloomberg and some are invalid such as Yahoo.finance.
Meanwhile, I thought that both Coinbase and Coingecko were used by UMA, it would be easier to use them as data sources.

smb2796 replied:
Usually Coinbase is used for crypto pairs that have an existing market on coinbase. When a market exists, it is easy to query for these prices through sources like Cryptowatch. Potentially querying for CNYUSD using a tool like TraderMade could work?

jishankai replied:
@smb2796 I tried TradeMade last week. How many requests should be satisfied a month if we use it as a source? Because it is charged, I need to figure out how much I should pay.  Besides, as PR #102 said it doesn't work on weekends, is it OK?

I referred to the UMIP-21 and cache the historical data from Coinbase & Coingecko by ourselves and build an endpoint for searching: [usdccny-api](https://github.com/jishankai/usdccny-api). 

Which plan is better? Must we support a forex-based feed?

smb2796 reviewed (commented):
Hey Shankai. Great UMIP, I'm really excited about the possibilities of creating FX derivatives!

Left a few comments - was generally curious about some of the extended rationale, and also think that the implementation/data sources could use some clarity.

One potential solution to clarify the price calculation methodology would be to build an example script that voters could use to query for prices. An example of this would be the [Cryptowatch price feed](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js) that was used as a reference implementation in UMIPs 6 & 7.

jishankai commented on UMIPs/umip-28.md:
```diff
@@ -0,0 +1,74 @@
+## Rationale
+Prices are primarily used by Priceless contracts to calculate a synthetic token’s redemptive value in case of liquidation or expiration. Contract counterparties also use the price index to ensure that sponsors are adequately collateralized. 
+
+Coinbase is a top exchange of the world. It is used by most of the price identifieres of UMA. Luckily, we can get USD(T)/CNY price by its api and it will be easy to add this price identifier.
```
No volume. There is no CNY token because it's illegal according to PRC laws. We can only exchange between USDT and CNY on the OTC desk of Binance, OKCoin, Huobi, Gate.io, etc.
We can find USDT history price in CNY unit by using Coinbase's API.

claytonroche replied:
https://www.loom.com/share/2f2eade930474900902a1ebd160530d1

^ Trying this out since there's a "Record a Loom" button!

smb2796 replied:
+1 Clayton for the new communication medium. I generally agree with the sentiment you expressed around using exchanges that trade the pair. If there is a sufficient reason to use crypto exchanges, I could potentially see using it - but without trading volume it seems like we be exposing ourselves to the validity of CB's secondhand reporting.

Does CB have a methodology published for how they determine the CNY/USDC price?

jishankai replied:
@claytonroche Got it. I am working on finding the root of it. It looks that the implementation of now doesn't satisfy the requirements.
I'll try finding firsthand and granular enough endpoints. 
@smb2796 No, CB doesn't.

jishankai commented on UMIPs/umip-28.md:
```diff
@@ -0,0 +1,74 @@
+
+| Endpoint | Field name of price |
+|------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| https://api.coinbase.com/v2/prices/USD-CNY/spot | *amount* |
```
Sorry for the mistake. I found that we couldn't get USDCNY historical price by using this endpoint, neither USDTCNY. We can only get USDCCNY historical price: 
`curl https://api.coinbase.com/v2/prices/USDC-CNY/spot\?date\=2020-11-30  -H 'Authorization: Bearer abd90df5f27a7b170cd775abf89d632b350b7c1c9d53e08b340cd9832ce52c2c'`
Maybe I should use USDCCNY price standing for USDCNY.

smb2796 replied:
If we still move forward with Coinbase as a source, yes this would be useful. The main issue I see with this is that it looks like it is only available on a daily granularity, which wouldn't be adequate for price determination

jishankai commented on UMIPs/umip-28.md:
```diff
@@ -0,0 +1,74 @@
+| Endpoint | Field name of price |
+|------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| https://api.coinbase.com/v2/prices/USD-CNY/spot | *amount* |
+| https://api.coingecko.com/api/v3/coins/tether/history | *current_price* |
```
Yes, it should be `https://api.coingecko.com/api/v3/coins/tether/history?date=30-12-2017`, I referred to UMIP-19 and didn't add the param in the table. 

We can get the current price by using this endpoint:
`https://api.coingecko.com/api/v3/simple/price?ids=tether&vs_currencies=cny`

I can't find a more granular basis. Why is a daily granularity basis not impractical? I think a current price for investors and a current price and a daily price for voters are OK,   aren't they?

I'll update the exact fields for votes.

claytonroche replied:
Daily is impractical because forex prices change constantly when the markets are open.  You could imagine a position being collateralized only 1% in excess of the requirement, and then prices could change a lot over a single day, and then when the price updates, suddenly it is either undercollateralized or or perhaps not attractive enough to liquidate.

Another possibility is that this position would be liquidated in the middle of the day if a liquidator reacted to the real market price, but since the UMIP is pointing towards this daily granularity, there is going to be ambiguity over which source to trust.  This could create unnecessary disputes / noise -- I believe that tokenholders would seek out higher granularity if the situation required it, but we should instead try to embed that same granularity in the price ID design.

jishankai replied:
Got it! I need to improve this. Can I refer to the UMIP-21 and cache the historical data by ourselves and build an endpoint for searching?

smb2796 replied:
That is a potential solution! The issue there is trust and creating a single point of failure with a caching solution. If there is other more granular historical data available, it would likely be optimal to not use a caching solution - but as a last resort it could potentially work given that voters could always use other less granular sources as a reference point.

smb2796 commented on UMIPs/umip-29.md:
````diff
@@ -0,0 +1,174 @@
+
+```
+
+### Coingecko Historical Data API
````
Awesome stuff explaining these endpoints. Thanks for adding clarity here.

nicholaspai commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,174 @@
+The definition of these identifiers should be:
+
+-----------------------------------------
+- Identifier names: **CNYUSD**
```
Why do you think it's necessary to have both a USDCNY and a CNYUSD identifier? Isn't one sufficient to cover all possible use cases?

jishankai replied:
It's not necessary to have both. The reason that I defined two was that I studied other UMIPs and each pair had two identifiers. It's more simple and clear if there is only one.

nicholaspai commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,174 @@
+-----------------------------------------
+- Identifier names: **CNYUSD**
+- Base Currency: CNY
+- Quote Currency: USD(C)
```
Isn't the quote currency USD because the identifier is tracking the CNY USD exchange rate not the CNY USDC rate?

jishankai replied:
The reason that I used USDC was that I could only get CNYUSDC price from Coinbase & Coingecko API. If we use forex API, the quote can be changed to USD.

nicholaspai commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,174 @@
+- Price Steps: 0.01 (2 decimals in more general trading format)
+-----------------------------------------
+
+- Exchanges: Coinbase, Coingecko.
```
Are crypto exchange APIs the best data providers of CNY USD? I would have thought it'd be better to use APIs like [one](https://exchangeratesapi.io/) which serves ECB data.

claytonroche replied:
Interesting - that one you linked @nicholaspai -- It's free?  And are you suggesting that we might form a price ID on taking the average of multiple APIs?  Such as https://exchangeratesapi.io/ AND tradermade?

jishankai replied:
Great!  I checked it and it is free. However, we can only get historical data by day on https://exchangeratesapi.io and it's impractical.

nicholaspai commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,174 @@
+## Rationale
+Prices are primarily used by Priceless contracts to calculate a synthetic token’s redemptive value in case of liquidation or expiration. Contract counterparties also use the price index to ensure that sponsors are adequately collateralized. 
+
+Coinbase is a top exchange of the world. It is used by most of the price identifieres of UMA. Luckily, we can get USD(C)/CNY price by its api and it will be easy to add this price identifier.
```
We are not restricted to using crypto exchanges as data providers, as explained in an earlier comment. I think we will get the most reliable data if we use several FX APIs that aren't necessarily focused on crypto data.

jishankai replied:
Got it. I have tried to find free or cheap FX APIs with better granularity data but I failed. Any recommendations would be appreciated.

nicholaspai reviewed (changes requested):
I think the USDCNY identifier will enable some really useful products that require access to the [USDCNY exchange rate](https://www.bloomberg.com/quote/USDCNY:CUR)! I have two main sources of feedback:

1) I don't think it makes sense to have both a USDCNY and CNYUSD identifier
2) I don't think we should constrain our choice of price resolution sources to crypto exchanges 

Overall this is a really well thought out UMIP and I'm looking forward to helping the proposer refine it.

jishankai commented:
Following the suggestions, 
- I removed CNYUSD and left one identifier.
- I updated USDCNY to USDCCNY which is more right here. If we use forex data sources, I'll change it back to USDCNY.
- I also set up the API server for caching the historical data by minutes.
- Please give me your feedback about how we should use TraderMade and ExchangeRatesAPI.

smb2796 reviewed (commented):
Awesome work on the caching solution. I think this could be presented as a backup data solution if you wish to support. 

TraderMade is not free, but does have adequate CNYUSD data and has been used as a data source in previous price identifiers.

Looks like TraderMade has CNYUSD historical data available with a 10 minute granularity. Example endpoint here: https://marketdata.tradermade.com/api/v1/minute_historical?currency=CNYUSD&date_time=2020-11-11-13:50&api_key=apikey. From briefly playing around with this, it should be noted that requests that do not fall on the 10 minute interval do not return data. There is also an hourly granularity endpoint here: https://marketdata.tradermade.com/api/v1/hour_historical?currency=EURUSD&date_time=2019-10-10-13:00&api_key=apikey.

TraderMade also has a live-data endpoint for USDCNY here: https://marketdata.tradermade.com/api/v1/live?currency=USDCNY&api_key=apikey

My suggestion would be to play around with TraderMade and edit the UMIP to use this.

jishankai commented:
> Awesome work on the caching solution. I think this could be presented as a backup data solution if you wish to support.
> 
> TraderMade is not free, but does have adequate CNYUSD data and has been used as a data source in previous price identifiers.
> 
> Looks like TraderMade has CNYUSD historical data available with a 10 minute granularity. Example endpoint here: https://marketdata.tradermade.com/api/v1/minute_historical?currency=CNYUSD&date_time=2020-11-11-13:50&api_key=apikey. From briefly playing around with this, it should be noted that requests that do not fall on the 10 minute interval do not return data. There is also an hourly granularity endpoint here: https://marketdata.tradermade.com/api/v1/hour_historical?currency=EURUSD&date_time=2019-10-10-13:00&api_key=apikey.
> 
> TraderMade also has a live-data endpoint for USDCNY here: https://marketdata.tradermade.com/api/v1/live?currency=USDCNY&api_key=apikey
> 
> My suggestion would be to play around with TraderMade and edit the UMIP to use this.

Thanks a lot for so many details. I'll update the UMIP.

smb2796 commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,123 @@
+### Weekend Price
+Over the weekend the REST API does not return any price, but we can request the price of a certain moment before the market close (as ex: the closing price of Friday).
+
+Due to unavailability of price feed for any financial asset over the weekend, tokenholders and users will be using the latest known price, before the market close on Friday, which esentially is the closing price of the asset from Friday. Same goes in a case of a liquidation process - the liquidator should use the last price from Friday in order to match with the price on which a synthetic asset w ...
```
Small nitpick, but more specifically a lack of availability for FX here. 

```suggestion
Due to unavailability of price feed for foreign exchange rates over the weekend, tokenholders and users will be using the latest known price, before the market close on Friday, which essentially is the closing price of the asset from Friday. Same goes in a case of a liquidation process - the liquidator should use the last price from Friday in order to match with the price on which a synthetic asset was created, if it was created over the weekend. If not the closing price on Friday for a certain asset should be a navigating point in calculating the collateralization ratio of a position and in the liquidation process.
```

smb2796 commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,123 @@
+
+If a request timestamp takes place on a weekend or any other day the Forex market is closed, voters should use the latest tick as the price. For the weekend that would be the closing price of the asset on Friday.
+
+
```
Pulling language from EURUSD UMIP draft to clarify that TraderMade is an example source.

```suggestion
TraderMade is provided as an accessible source to query for this data, but ultimately how one queries for these rates should be varied and determined by the voter to ensure that there is no central point of failure.
```

smb2796 commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,123 @@
+
+## Implementation
+
+### TraderMade Historical API
```
We need to pick one of these and match the `Pricing Interval` to the granularity of the data. I recommend going with the 10 minute granularity and providing instructions on how voters should round time down to the closest 10 minute interval.

smb2796 commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,123 @@
+If a request timestamp takes place on a weekend or any other day the Forex market is closed, voters should use the latest tick as the price. For the weekend that would be the closing price of the asset on Friday.
+
+
+## Security Considerations
```
Recommend some language on the security considerations around not having data available over the weekend/certain holidays. https://github.com/UMAprotocol/UMIPs/pull/102/files should be a good reference

smb2796 commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,123 @@
+- Base Currency: CNY
+- Quote Currency: USD
+- Result Processing: None
+- Price Steps: 0.0001 (4 decimals in more general trading format)
```
Is the intention to use this price id with USDC as collateral? If so, this should be 6 decimals to match the precision of USDC

claytonroche commented on UMIPs/umip-29.md:
````diff
@@ -0,0 +1,120 @@
+}
+```
+
+From briefly playing around with this, it should be noted that requests that do not fall on the 10 minute interval do not return data. Voters should round time down or up to the closest 10 minute interval if they request the price which doesn't fall on the 10 minute interval.
````
I would remove the language "from briefly playing around with this"

The rounding solution seems fitting to me, but I think you should change the "pricing interval" in the tech specs to 10 minutes then.

smb2796 commented on UMIPs/umip-29.md:
```diff
@@ -0,0 +1,120 @@
+Tradermade’s price feed is an aggregated feed from multiple Tier One and Two Banks, Market-Makers and Data Providers. They are popular with Quantitative Traders, Fintech companies and Institutional customers who require a high quality and trusted feed.
+
+
+## Implementation
```
The resources you provided are good, but this section should also give the steps that a voter should take to resolve a price. What endpoint do they query, how do they determine the time to query it at, what do they do after getting the price back

Tommy1231232 commented on UMIPs/umip-32.md:
```diff
@@ -0,0 +1,126 @@
+| Created    | December 16, 2020                                                                                                                           |
+
+## Summary (2-5 sentences)
+The DVM should support price requests for the CNY/USD price index. 
```
Is this for the on-shore or off-shore price of CNYUSD? What does Tradermade use? As it is written now it looks like it is the on-shore rate. Do you know if tradermade is reporting the on-shore or off-shore rate?

https://statrys.com/blog/cnh-vs-cny-differences-chinese-renminbi

jishankai replied:
This is for on-shore price. Both on-shore and off-shore price are supported by TraderMade. The currency id of off-shore price is CNH.

On-shore endpoint:
```
https://marketdata.tradermade.com/api/v1/live?currency=CNYUSD&api_key={APIKEY}
```
Off-shore endpoint:
```
https://marketdata.tradermade.com/api/v1/live?currency=CNHUSD&api_key={APIKEY}
```

smb2796 commented on UMIPs/umip-32.md:
```diff
@@ -0,0 +1,126 @@
+## Motivation
+The DVM currently does not support the CNY/USD index.
+
+Supporting the CNYUSD price identifier would enable the creation of a Chinese Yuan FX derivative, backed by USD. Token minters could go short on the CNY/USD index, while token holders could go long or use synthetic fxCNY for functional purposes.
```
Just to confirm - is your intention to use this price id with USDC as collateral?

jishankai replied:
I think the collateral can be USDT, USDC, or DAI. I haven't decided. What does this matter?

smb2796 replied:
For now this matters because price identifiers are returned with a specific amount of decimals, so we want this to match up with the collateral that you plan to use. If you deployed two contracts using say DAI and USDC, one contract would not behave correctly with the returned price because values in DAI are returned as 1e18, while values in USDC are returned as 1e6. I put a recommendation in the technical specifications section to be explicit about the collateral decimals that are intended to be used.

smb2796 commented on UMIPs/umip-32.md:
````diff
@@ -0,0 +1,126 @@
+```
+https://marketdata.tradermade.com/api/v1/minute_historical?currency=CNYUSD&date_time=2020-11-11-13:50&api_key={apikey}
+```
+and use the close price as reference. Because we use the endpoint with a 10 minute granularity, to clarify if the price request timestamp is as ex 13:01:59, voters should query 13:00 as a timestamp. If the price request timestamp is as ex 13:06:59, votes should query 13:10 as a timestamp. 
````
I think it's problematic to have voters round timestamps up as that price does not "exist" yet when a price has been requested. I'm in favor of rounding all timestamps down, no matter where this falls on the 10 minute interval

jishankai replied:
Thanks. Updated it.

smb2796 commented on UMIPs/umip-32.md:
````diff
@@ -0,0 +1,126 @@
+  "request_time": "Tue, 29 Dec 2020 09:02:04 GMT"
+}
+```
+
````
`Voters should use the `close` price and round to 6 decimals.`

smb2796 commented on UMIPs/umip-32.md:
```diff
@@ -0,0 +1,122 @@
+- Identifier names: **CNYUSD**
+- Base Currency: CNY
+- Quote Currency: USD
+- Result Processing: None
```
We should pick an initial collateral type to use for the reasons I explained above and be explicit about the collateral decimals. If you wanted to initially use USDC or USDT, this would be 6. Dai would be 18.

```suggestion
- Result Processing: None
- Collateral Decimals: 18
```

smb2796 replied:
Going to commit as 18 decimals (using DAI) for now. Please submit another pr if you would like this to be changed.

smb2796 reviewed (commented):
Looking great. Once you address the collateral question I think we can merge into `Last Call` status

smb2796 reviewed (approved):
Approving into Last Call based on if the collateral type was 18 decimals (DAI). Please let me know or submit a pull request if you would like this detail changed.
